### PR TITLE
[Driver] Add toolchain for X86_64 UEFI target

### PR DESCRIPTION
**Pull request by Prabhuk**
Introduce changes necessary for UEFI X86_64 target Clang driver.
Addressed the review comments originally suggested in Phabricator.

Differential Revision: https://reviews.llvm.org/D159541

### CODE REVIEW

File: clang/lib/Driver/ToolChains/UEFI.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+Tool *UEFI::buildLinker() const { return new tools::uefi::Linker(*this); }
+
+void tools::uefi::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+                                       const InputInfo &Output,
+                                       const InputInfoList &Inputs,
+                                       const ArgList &Args,
+                                       const char *LinkingOutput) const {
+  ArgStringList CmdArgs;
+
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.cpp:48]: delete line

File: clang/lib/Driver/ToolChains/UEFI.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  CmdArgs.push_back("-nologo");
+
+  // TODO: Other UEFI binary subsystems that are currently unsupported:
+  // efi_boot_service_driver, efi_rom, efi_runtime_driver.
+  CmdArgs.push_back("-subsystem:efi_application");
+
+  // Default entry function name according to the TianoCore reference
+  // implementation is EfiMain.
+  // TODO: Provide a flag to override the entry function name.
+  CmdArgs.push_back("-entry:EfiMain");
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.cpp:65]: Is there a specification that this is `EfiMain`?

File: clang/lib/Driver/ToolChains/UEFI.cpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    // or -L.
+    A.renderAsInput(Args, CmdArgs);
+  }
+
+  // This should ideally be handled by ToolChain::GetLinkerPath but we need
+  // to special case some linker paths. In the case of lld, we need to
+  // translate 'lld' into 'lld-link'.
+  StringRef Linker =
+      Args.getLastArgValue(options::OPT_fuse_ld_EQ, CLANG_DEFAULT_LINKER);
+  if (Linker.empty() || Linker.equals_insensitive("lld"))
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.cpp:107]: If we don't model after existing constraints (e.g. MSVC), avoid `_insensitive`

File: clang/lib/Driver/ToolChains/UEFI.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  CmdArgs.push_back("-entry:EfiMain");
+
+  // "Terminal Service Aware" flag is not needed for UEFI applications.
+  CmdArgs.push_back("-tsaware:no");
+
+  // EFI_APPLICATION to be linked as DLL by default.
+  CmdArgs.push_back("-dll");
+
+  if (Args.hasArg(options::OPT_g_Group, options::OPT__SLASH_Z7))
+    CmdArgs.push_back("-debug");
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.cpp:74]: untested

File: clang/lib/Driver/ToolChains/UEFI.cpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    if (Input.isFilename()) {
+      CmdArgs.push_back(Input.getFilename());
+      continue;
+    }
+
+    const Arg &A = Input.getInputArg();
+    if (A.getOption().matches(options::OPT_l)) {
+      StringRef Lib = A.getValue();
+      const char *LinkLibArg;
+      if (Lib.ends_with(".lib"))
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.cpp:89]: The `-l` behavior is untested. Does it really behave this way detecting `.lib`? lld-link appends `.lib` if the name does not contain a `.`.

For linker input, `AddLinkerInputs(ToolChain, Inputs, Args, CmdArgs, JA);`

File: clang/lib/Driver/ToolChains/UEFI.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+protected:
+  Tool *buildLinker() const override;
+
+public:
+  bool HasNativeLLVMSupport() const override { return true; }
+  UnwindTableLevel
+  getDefaultUnwindTableLevel(const llvm::opt::ArgList &Args) const override {
+    return UnwindTableLevel::Asynchronous;
+  }
+  bool isPICDefault() const override { return true; }
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.h:50]: Q: why is this PIC and forced (`-fno-pic` will error)?

File: clang/lib/Basic/Targets/X86.h [hunk 7]
```diff
@@ -819,6 +819,43 @@ class LLVM_LIBRARY_VISIBILITY X86_64TargetInfo : public X86TargetInfo {
   }
 };
 
+// x86-64 UEFI target
+class LLVM_LIBRARY_VISIBILITY UEFIX86_64TargetInfo
+    : public UEFITargetInfo<X86_64TargetInfo> {
+public:
+  UEFIX86_64TargetInfo(const llvm::Triple &Triple, const TargetOptions &Opts)
+      : UEFITargetInfo<X86_64TargetInfo>(Triple, Opts) {
```
Comment (MaskRay) [inline clang/lib/Basic/Targets/X86.h:827]: This clangBasic change needs a `clang/test/CodeGen` test for things including: data layout , calling convention.

**Review by MaskRay — CHANGES_REQUESTED** [review]

**Comment by brad0** [comment]
Ping.

File: clang/lib/Basic/Targets/X86.h (hunk 7 above)
Comment (Prabhuk) [reply clang/lib/Basic/Targets/X86.h:833]: I have added layout test to target-data.c. I am trying to figure out where to add a test for the calling convention.

File: clang/lib/Driver/ToolChains/UEFI.cpp (hunk 2 above)
Comment (Prabhuk) [reply clang/lib/Driver/ToolChains/UEFI.cpp:64]: For the initial implementation to compile simple EFI applciations, I am using Tianacore's reference implementation as mentioned in the inline comment. I had mentioned this in my RFC here: https://discourse.llvm.org/t/rfc-uefi-driver-support-uefi-target/73261

File: clang/lib/Driver/ToolChains/UEFI.h (hunk 6 above)
Comment (Prabhuk) [reply clang/lib/Driver/ToolChains/UEFI.h:50]: I am matching the behavior of the MSVC toolchain in setting isPICDefault and isPICDefaultForced.

File: clang/lib/Driver/ToolChains/UEFI.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  Args.AddAllArgValues(CmdArgs, options::OPT__SLASH_link);
+
+  AddLinkerInputs(TC, Inputs, Args, CmdArgs, JA);
+
+  // This should ideally be handled by ToolChain::GetLinkerPath but we need
+  // to special case some linker paths. In the case of lld, we need to
+  // translate 'lld' into 'lld-link'.
+  StringRef Linker =
+      Args.getLastArgValue(options::OPT_fuse_ld_EQ, CLANG_DEFAULT_LINKER);
+  if (Linker.empty() || Linker.equals("lld"))
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.cpp:84]: Nit: Prefer `==` to `equals`

File: clang/lib/Driver/ToolChains/UEFI.cpp [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    Linker = "lld-link";
+
+  auto LinkerPath = TC.GetProgramPath(Linker.str().c_str());
+  auto LinkCmd = std::make_unique<Command>(
+      JA, *this, ResponseFileSupport::AtFileUTF16(),
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.cpp:89]: Is this intended to be UTF-16 like MSVC?

File: clang/lib/Driver/ToolChains/UEFI.cpp [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+using namespace clang::driver;
+using namespace clang::driver::toolchains;
+using namespace clang;
+using namespace llvm::opt;
+
+UEFI::UEFI(const Driver &D, const llvm::Triple &Triple, const ArgList &Args)
+    : ToolChain(D, Triple, Args) {
+  getProgramPaths().push_back(getDriver().getInstalledDir());
+  if (getDriver().getInstalledDir() != getDriver().Dir)
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.cpp:36]: I have removed getInstalledDir in #80527 and follow-ups. This needs an update.

File: clang/lib/Driver/ToolChains/UEFI.cpp (hunk 4 above)
Comment (MaskRay) [reply clang/lib/Driver/ToolChains/UEFI.cpp:73]: This logic from MSVC.cpp seems incorrect to me. If the intention is for  `-g` to enable `-debug`, then `-g -g0` should be able to remove `-debug`.

File: clang/lib/Driver/ToolChains/UEFI.cpp [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  // "Terminal Service Aware" flag is not needed for UEFI applications.
+  CmdArgs.push_back("-tsaware:no");
+
+  // EFI_APPLICATION to be linked as DLL by default.
+  CmdArgs.push_back("-dll");
+
+  if (Args.hasArg(options::OPT_g_Group, options::OPT__SLASH_Z7))
+    CmdArgs.push_back("-debug");
+
+  Args.AddAllArgValues(CmdArgs, options::OPT__SLASH_link);
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.cpp:75]: We need tests for this.

File: clang/lib/Driver/ToolChains/UEFI.cpp (hunk 10 above)
Comment (Prabhuk) [reply clang/lib/Driver/ToolChains/UEFI.cpp:36]: I see the getInstalledDir still present after your changes. Except the definition been changed. Am I missing something here?

https://github.com/llvm/llvm-project/commit/2b5cd8be3af43e5aa5b76b6aeb1edd3141b803ca#diff-2c12c1bdb5245225094e1125ceb698aa4ed7e69cd1d154a4004f5adc54691624R428

File: clang/lib/Driver/ToolChains/UEFI.cpp (hunk 9 above)
Comment (Prabhuk) [reply clang/lib/Driver/ToolChains/UEFI.cpp:89]: Yes. This maintains compatibility with LINK.exe which only supports UTF-16.

**Comment by MaskRay** [comment]
Some older toolchains were probably contributed with a lot of `CmdArgs.push_back` uncovered by tests. They are not good examples to follow. For new toolchains, we do want all constructed `CmdArgs.push_back` to be covered. This allows refactoring by someone who is unfamiliar with your usage pattern.

File: clang/lib/Basic/Targets/X86.h (hunk 7 above)
Comment (MaskRay) [reply clang/lib/Basic/Targets/X86.h:833]: Still waiting for a calling convention test

Comment (Prabhuk) [reply clang/lib/Basic/Targets/X86.h:822]: Please see `clang/Driver/unittests/ToolchainTest.cpp`

https://github.com/llvm/llvm-project/pull/76838/files#diff-2a39548e3b4a62a12affcb6dbe9e8f563cddc2aa50aa7e24b94a1de0bb91be1eR578

File: clang/lib/Driver/ToolChains/UEFI.cpp (hunk 8 above)
Comment (Prabhuk) [reply clang/lib/Driver/ToolChains/UEFI.cpp]: Done.

File: clang/lib/Driver/ToolChains/UEFI.cpp (hunk 10 above)
Comment (Prabhuk) [reply clang/lib/Driver/ToolChains/UEFI.cpp]: Made changes to reflect removal of getInstalledDir.

File: clang/lib/Driver/ToolChains/UEFI.cpp (hunk 11 above)
Comment (Prabhuk) [reply clang/lib/Driver/ToolChains/UEFI.cpp:71]: Please see `clang/test/Driver/uefi-constructed-args.c`.

**Comment by Prabhuk** [comment]
> Some older toolchains were probably contributed with a lot of `CmdArgs.push_back` uncovered by tests. They are not good examples to follow. For new toolchains, we do want all constructed `CmdArgs.push_back` to be covered. This allows refactoring by someone who is unfamiliar with your usage pattern.

Thank you! I believe the test in `clang/test/Driver/uefi-constructed-args.c` covers all constructed args. PTAL.

**Review by mysterymath — COMMENTED** [review]
LGTM from my POV.

File: clang/test/Driver/uefi-data-layout.c [hunk 12]
```diff
@@ -0,0 +1,3 @@
+// RUN: %clang -target x86_64-unknown-uefi -S -emit-llvm -o - %s | \
+// RUN:     FileCheck --check-prefix=X86_64_UEFI %s
+// X86_64_UEFI: target datalayout = "e-m:w-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
```
[the remark this answers is not in the document]
Comment (Prabhuk) [reply clang/test/Driver/uefi-data-layout.c]: Thank you! Added end of file new lines for the new test files.

File: clang/lib/Driver/ToolChains/UEFI.cpp [hunk 13]
```diff
@@ -0,0 +1,88 @@
+//===--- UEFI.h - UEFI ToolChain Implementations ----------*- C++ -*-===//
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.cpp:1]: first line is incorrect. ensure 80-char wide and omit `C++` for `.cpp` files.

File: clang/lib/Driver/ToolChains/UEFI.h [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_LIB_DRIVER_TOOLCHAINS_UEFI_H
+#define LLVM_CLANG_LIB_DRIVER_TOOLCHAINS_UEFI_H
+
+#include "clang/Driver/Tool.h"
+#include "clang/Driver/ToolChain.h"
+
+namespace clang {
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/UEFI.h:15]: `namespace clang::driver {`

File: clang/test/Driver/uefi-data-layout.c [hunk 15]
```diff
@@ -0,0 +1,4 @@
+// RUN: %clang -target x86_64-unknown-uefi -S -emit-llvm -o - %s | \
```
Comment (MaskRay) [inline clang/test/Driver/uefi-data-layout.c:1]: test/Driver tests should generally only test `-###`, not `-emit-llvm -S` (codegen). Place such codegen tests to test/CodeGen

File: clang/test/Driver/uefi-data-layout.c [hunk 16]
```diff
@@ -0,0 +1,4 @@
...
+// RUN:     FileCheck --check-prefix=X86_64_UEFI %s
+// X86_64_UEFI: target datalayout = "e-m:w-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
+
```
Comment (MaskRay) [inline clang/test/Driver/uefi-data-layout.c:4]: delete trailing blank line

**Comment by MaskRay** [comment]
`[Driver] Add toolchain for ...' might be a more conventional patch subject.

File: clang/test/Driver/uefi-constructed-args.c [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+// CHECK-SAME: "-triple" "x86_64-unknown-uefi"
+// CHECK-SAME: "-mrelocation-model" "pic" "-pic-level" "2"
+// CHECK-SAME: "-mframe-pointer=all"
+// CHECK-NEXT: "-nologo"
+// CHECK-SAME: "-subsystem:efi_application"
+// CHECK-SAME: "-entry:EfiMain"
+// CHECK-SAME: "-tsaware:no"
+// CHECK-SAME: "-dll"
+// CHECK-SAME: "-debug"
+
```
Comment (MaskRay) [inline clang/test/Driver/uefi-constructed-args.c:14]: delete trailing blank line

File: clang/test/CodeGen/target-data.c [hunk 18]
```diff
@@ -22,6 +22,10 @@
 // RUN:     FileCheck --check-prefix=X86_64 %s
 // X86_64: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
 
+// RUN: %clang_cc1 -triple x86_64-unknown-uefi -emit-llvm -o - %s | \
```
Comment (MaskRay) [inline clang/test/CodeGen/target-data.c:25]: With clang/test/CodeGen/X86/uefi-data-layout.c, this is unneeded

**Review by MaskRay — APPROVED** [review]

**Comment by bhandarkar-pranav** [comment]
Hi @Prabhuk, 
I am seeing this issue https://github.com/llvm/llvm-project/issues/109328 after this PR. Here is a proposed [fix](https://github.com/llvm/llvm-project/pull/109329). Could you please check?

**Comment by nico** [comment]
Looks like this breaks tests on Mac: http://45.33.8.238/macm1/92471/step_6.txt

Please take a look and revert for now if it takes a while to fix.

**Comment by bhandarkar-pranav** [comment]
> Looks like this breaks tests on Mac: http://45.33.8.238/macm1/92471/step_6.txt
> 
> Please take a look and revert for now if it takes a while to fix.
@nico 
I cannot access the link from my work VPN, but if this is the issue (https://github.com/llvm/llvm-project/issues/109328) then here is a fix (https://github.com/llvm/llvm-project/pull/109329)

**Comment by Prabhuk** [comment]
I am going to revert this and check the failures after.

**Pull request merged**